echo/app/src/main/jni/audio_main.cpp

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2015 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cassert>
+#include <cstring>
+#include <jni.h>
+
+#include <sys/types.h>
+#include <SLES/OpenSLES.h>
+
+#include "audio_common.h"
+#include "audio_recorder.h"
+#include "audio_player.h"
+
+struct EchoAudioEngine {
+    SLmilliHertz fastPathSampleRate_;
+    uint32_t     fastPathFramesPerBuf_;
+    uint16_t     sampleChannels_;
+    uint16_t     bitsPerSample_;
+
+    SLObjectItf  slEngineObj_;
+    SLEngineItf  slEngineItf_;
+
+    AudioRecorder  *recorder_;
+    AudioPlayer    *player_;
+    AudioQueue     *freeBufQueue_;    //Owner of the queue
+    AudioQueue     *recBufQueue_;     //Owner of the queue
+
+    sample_buf  *bufs_;
+    uint32_t     bufCount_;
+    uint32_t     frameCount_;
+};
+static EchoAudioEngine engine;
+
+bool EngineService(void* ctx, uint32_t msg, void* data );
+
+extern "C" {
+JNIEXPORT void JNICALL
+        Java_com_google_sample_echo_MainActivity_createSLEngine(JNIEnv *env, jclass, jint, jint);
+JNIEXPORT void JNICALL
+        Java_com_google_sample_echo_MainActivity_deleteSLEngine(JNIEnv *env, jclass type);
+JNIEXPORT jboolean JNICALL
+        Java_com_google_sample_echo_MainActivity_createSLBufferQueueAudioPlayer(JNIEnv *env, jclass);
+JNIEXPORT void JNICALL
+        Java_com_google_sample_echo_MainActivity_deleteSLBufferQueueAudioPlayer(JNIEnv *env, jclass type);
+
+JNIEXPORT jboolean JNICALL
+        Java_com_google_sample_echo_MainActivity_createAudioRecorder(JNIEnv *env, jclass type);
+JNIEXPORT void JNICALL
+        Java_com_google_sample_echo_MainActivity_deleteAudioRecorder(JNIEnv *env, jclass type);
+JNIEXPORT void JNICALL
+        Java_com_google_sample_echo_MainActivity_startPlay(JNIEnv *env, jclass type);
+JNIEXPORT void JNICALL
+        Java_com_google_sample_echo_MainActivity_stopPlay(JNIEnv *env, jclass type);
+}
+
+JNIEXPORT void JNICALL
+Java_com_google_sample_echo_MainActivity_createSLEngine(
+        JNIEnv *env, jclass type, jint sampleRate, jint framesPerBuf) {
```

Comment:
maybe you could add those as method of the EchoAudioEngine type? and keep the JNI function as a wrapper (get the instance call the corresponding method).